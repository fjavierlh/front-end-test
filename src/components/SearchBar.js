import React, { useEffect, useState } from 'react';
import { Hint } from 'react-autocomplete-hint';
import styled from 'styled-components';
import breakpoints from '../constants/devices-sizes';

const SearchBar = ({ onSearch, products }) => {
  const [value, setValue] = useState('');
  const [autocompleteSuggestions, setAutocompleteSuggestions] = useState([]);

  useEffect(() => {
    const suggestionsRaw =
      products.length !== 0
        ? products
            .map(({ brand, model }) => [
              `${brand} ${model}`,
              `${model} ${brand}`,
              `${brand}`,
              `${model}`,
            ])
            .reduce((acc, curr) => acc.concat(curr))
        : [];
    const suggestions = [...new Set(suggestionsRaw)];
    setAutocompleteSuggestions(suggestions);
  }, [products]);

  const onChange = (e) => {
    const value = e.target.value;
    setValue(value);
    onSearch(value.toLowerCase());
  };

  return (
    <SearchBarStyled>
      <Hint className="hint" options={autocompleteSuggestions} allowTabFill>
        <input
          type="text"
          placeholder="Search model or brand..."
          onChange={onChange}
          value={value}
        />
      </Hint>
    </SearchBarStyled>
  );
};

const SearchBarStyled = styled.div`
  & * input {
    width: 100%;
    padding: 0.5em;
    border: none;
    font-size: 1.5rem;
    border-radius: 5px;
    &::placeholder {
      animation: fadeIn 1s ease-in-out;
      color: #333;
      font-weight: bold;
    }

    &:focus {
      outline: none;

      &::placeholder {
        color: #ccc;
      }
    }
  }
  @media only screen and (${breakpoints.device.xs}) {
    flex-basis: 100%;
    margin: 1em 0 1em 0;
  }
  @media only screen and (${breakpoints.device.sm}) {
    margin: 1em 0 1em 50%;
  }
  @media only screen and (${breakpoints.device.lg}) {
  }
`;

export default SearchBar;
