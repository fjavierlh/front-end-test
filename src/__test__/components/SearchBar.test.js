import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { allProducts } from '../../../fixtures/allProducts.fixture';
import SearchBar from '../../components/SearchBar';

const onSearchMock = jest.fn();

describe('SearchBar', () => {
  test('should render SearchBar component', () => {
    const { container } = render(
      <SearchBar onSearch={onSearchMock} products={allProducts} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should render search input with placeholder text', () => {
    render(<SearchBar onSearch={onSearchMock} products={allProducts} />);
    expect(
      screen.queryByPlaceholderText(/search model or brand/i)
    ).toBeInTheDocument();
  });

  test('should render search input and allow user to type', () => {
    render(<SearchBar onSearch={onSearchMock} products={allProducts} />);
    const searchInput = screen.getByPlaceholderText(/search model or brand/i);
    expect(searchInput).toBeInTheDocument();

    act(() => {
      userEvent.click(searchInput);
    });

    expect(searchInput).toHaveFocus();
    expect(searchInput).toHaveValue('');

    userEvent.type(searchInput, 'test');
    expect(searchInput.value).toBe('test');
  });
});
