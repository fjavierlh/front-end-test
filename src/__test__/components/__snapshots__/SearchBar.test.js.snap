// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should render SearchBar component 1`] = `
<div>
  <div
    class="sc-bdvvtL fpEDfN"
  >
    <div
      class="rah-input-wrapper"
      style="position: relative;"
    >
      <input
        placeholder="Search model or brand..."
        style="box-sizing: border-box;"
        type="text"
        value=""
      />
      <span
        class="rah-hint-wrapper"
        style="display: flex; pointer-events: none; background-color: transparent; border-color: transparent; box-sizing: border-box; box-shadow: none; color: rgba(0, 0, 0, 0.35); position: absolute; top: 0px; left: 0px; font-family: -webkit-small-control; font-size: 1.5rem; width: 100%; line-height: normal; padding: 0.5em 0.5em 0.5em 0.5em; border-style: solid solid solid solid; border-width: 1px 1px 1px 1px;"
      >
        <span
          class="rah-text-filler"
          style="visibility: hidden; pointer-events: none; white-space: pre;"
        />
        <input
          class="rah-hint"
          style="pointer-events: none; background: transparent; width: 100%; outline: none; box-shadow: none; padding: 0px; margin: 0px; color: rgba(0, 0, 0, 0.3); caret-color: transparent; font-family: -webkit-small-control; font-size: 1.5rem; line-height: normal;"
          tabindex="-1"
          value=""
        />
      </span>
    </div>
  </div>
</div>
`;
